### PR TITLE
+ show the inferred exception list (as part of the type) for functions that don't have an explicit `.raises` pragma

### PR DESCRIPTION
**Pull request by nickysn**

### CODE REVIEW

**Comment by arnetheduck** [comment]
maybe introduce a syntax for inferred / augmented information, ie `{.*raises: [..]*, ....}` or something like that? 

that said, generally, inferred raises has less information value as it doesn't represent a deliberate choice by the author of the API, so it should also probably go behind everything else that was explicitly included in the signature (if a function has more pragmas)

**Comment by nickysn** [comment]
`.*raises: [..]*` looks way more cryptic than `.raises: <inferred> [..]`. Will people know what it means?

Moving it last, after the explicitly set pragmas is fine with me. I'll update the pull request.

**Comment by arnetheduck** [comment]
ideally it would appear in italic or some other font, which * is in markdown - not sure what the possibilities are ... ie type inlay hints appear greyed in rust in vscode for example.

**Comment by nickysn** [comment]
`.*raises: [..]*` looks way more cryptic than `.raises: <inferred> [..]`. Will people know what it means?

Moving it last, after the explicitly set pragmas is fine with me. I'll update the pull request.

> ideally it would appear in italic or some other font, which * is in markdown - not sure what the possibilities are ... ie type inlay hints appear greyed in rust in vscode for example.

It doesn't work like that. First, for inlay hints, we have no control over the color of the text, it is decided by the VS code color theme. The Nim inlay type hints and the Rust inlay type hints are exactly the same color, for this particular reason.

But this isn't shown in an inlay hint, this is shown on hover, and it is indeed colored, but it follows the colors for Nim syntax highlighting, and is not interpreted as markdown.

Here's how it would look with `.*raises: [..]*`:
image

And here's how it looks with `.raises: <inferred> [..]`:
image

This info will also be shown on the console as part of compiler error messages and the console also doesn't support markdown.

File: compiler/types.nim [hunk 1]
```diff
@@ -723,20 +723,35 @@ proc typeToString(typ: PType, prefer: TPreferedDesc = preferName): string =
...
           prag.add($raisesSpec)
-
+          hasImplicitRaises = true
       if tfNoSideEffect in t.flags:
         addSep(prag)
         prag.add("noSideEffect")
       if tfThread in t.flags:
         addSep(prag)
         prag.add("gcsafe")
+      if not hasImplicitRaises and not isNil(t.owner) and not isNil(t.owner.typ) and not isNil(t.owner.typ.n) and (t.owner.typ.n.len > 0):
```
Comment (Araq) [inline compiler/types.nim]: Needs to be opt-in via the `prefer` parameter as in general the error messages became more spammy and only nimsuggest wants it.

Comment (nickysn) [reply compiler/types.nim]: Ok, done. Introduced preferInferredEffects. The error messages are unchanged now, only nimsuggest hints use it.

**Pull request merged**